src/tir/schedule/concrete_schedule.cc

```diff
@@ -439,6 +439,44 @@ BlockRV ConcreteScheduleNode::CacheWrite(const BlockRV& block_rv, int write_buff
 
 /******** Schedule: Compute location ********/
 
+void ConcreteScheduleNode::ComputeAt(const BlockRV& block_rv, const LoopRV& loop_rv,
+                                     bool preserve_unit_loops) {
+  static StmtSRef inline_mark = StmtSRef::InlineMark();
+  static StmtSRef root_mark = StmtSRef::RootMark();
+  StmtSRef loop_sref = this->GetSRef(loop_rv);
```

Comment:
We should never ever insert `InlineMark` and `RootMark` into the symbol table, if we do, it is a bug. The two marks are only used to convert a compute-at into compute-inline, and thus is more like a constant, not random variables in the symbol table

Reply:
Hmmm.. If they are never inserted into the symbol table, how possibly could we find them in symbol table here? https://github.com/apache/tvm/blob/b01ab9e81e6a9605e6d2dce5b0c81ce551c1839b/src/tir/schedule/concrete_schedule.h#L210-L216

Reply:
Here is how it works:
- `SampleComputeLocation` returns `LoopRV`s that indicate which loops the current stage can be moved under. Besides, `SampleComputeLocation` provides options to "do nothing" and "do compute-inline" instead of doing compute-at;
- If the primitive wants to instruct the system to "do nothing" or "do compute-inline" instead, it returns LoopRVs that are looked up as inline-mark or root-mark;
- Therefore, on line 211 & 214 in GetSRef, we check if the loop_sref is inline-mark or root-mark.
- In compute-at, it checks the return value of `GetSRef` to determine which primitive it should dispatch to

> We should never ever insert InlineMark and RootMark into the symbol table, if we do, it is a bug

Sorry I didn't make it clear. I mean they shouldn't be in the symbol table as keys, because they are not loop_rvs. I made it wrong previously in our private repo so I remembered this bug quite vividly lol.

Reply:
Thanks for your thorough explanation! I can completely understand the workflow now 😆